lib/IRGen/GenDistributed.cpp

```diff
@@ -836,15 +838,35 @@ ArgumentDecoderInfo DistributedAccessor::findArgumentDecoder(
                        instance);
 
     decoder = instance.claimNext();
-  }
 
-  auto *decodeSIL = IGM.getSILModule().lookUpFunction(SILDeclRef(decodeFn));
-  auto *fnPtr = IGM.getAddrOfSILFunction(decodeSIL, NotForDefinition,
-                                         /*isDynamicallyReplaceable=*/false);
+    /// When using library evolution functions have another "dispatch thunk"
+    /// so we must use this instead of the decodeFn directly.
+   if (classDecl->hasResilientMetadata() && !classDecl->isFinal()) {
+      if (SILDeclRef(decodeFn).getOverriddenVTableEntry()) {
+        SILDeclRef(decodeFn).getOverriddenVTableEntry().getDecl()->getDeclContext()->dumpContext();
+        fnPtr = IGM.getAddrOfDispatchThunk(SILDeclRef(decodeFn), NotForDefinition);
+        usesDispatchThunk = true;
+      }
+    }
+  }
 
-  auto methodPtr = FunctionPointer::forDirect(
-    classifyFunctionPointerKind(decodeSIL), fnPtr,
-    /*secondaryValue=*/nullptr, signature);
+  if (!fnPtr) {
+    auto *decodeSIL = IGM.getSILModule().lookUpFunction(SILDeclRef(decodeFn));
+    fnPtr = IGM.getAddrOfSILFunction(decodeSIL, NotForDefinition,
+        /*isDynamicallyReplaceable=*/false);
+  }
+  assert(fnPtr);
+
+  FunctionPointer methodPtr;
+  if (usesDispatchThunk) {
```

Comment:
This can't be right. You're either using a dispatch thunk (which virtually dispatches the method) or directly calling the implementation (even though it may be overridden by a subclass). Don't you also have to handle a non-resilient vtable call?

Reply:
I found what SILGen is using - `getMethodDispatch` to check, we should do `getMethodDispatch(decodeFn) == MethodDispatch::Class)` and that should do the trick.

Reply:
Ah let’s try that, thank you!
So another test to add here is probably a decoder that uses inheritance then. Thanks folks!

Reply:
Yeah, that would be useful to have all of the combinations.

Reply:
Hah, we were not proberly handling ad-hoc impls in superclasses either tbh!

Now fixed both a using the getDispatchMethod -- thank you!